controllers/csiaddonsnode_controller.go

```diff
@@ -39,17 +50,69 @@ type CSIAddonsNodeReconciler struct {
 
 // Reconcile is part of the main kubernetes reconciliation loop which aims to
 // move the current state of the cluster closer to the desired state.
-// TODO(user): Modify the Reconcile function to compare the state specified by
-// the CSIAddonsNode object against the actual cluster state, and then
-// perform operations to make the cluster state reflect the state specified by
-// the user.
-//
-// For more details, check Reconcile and its Result here:
-// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.10.0/pkg/reconcile
 func (r *CSIAddonsNodeReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	logger := log.FromContext(ctx)
+
+	// Fetch CSIAddonsNode instance
+	csiAddonsNode := &csiaddonsv1alpha1.CSIAddonsNode{}
+	if err := r.Client.Get(ctx, req.NamespacedName, csiAddonsNode); err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			logger.Info("CSIAddonsNode resource not found")
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	// TODO: add validation for csiaddonsnode spec
+	nodeID := csiAddonsNode.Spec.Driver.NodeID
+	driverName := csiAddonsNode.Spec.Driver.Name
+	endPoint := csiAddonsNode.Spec.Driver.EndPoint
+	key := r.createKey(csiAddonsNode)
+	logger = logger.WithValues("NodeID", nodeID, "EndPoint", endPoint, "Driver Name", driverName)
+
+	if !csiAddonsNode.DeletionTimestamp.IsZero() {
+		// if deletion timestap is set, the CSIAddonsNode is getting deleted,
+		// delete connections and remove finalizer.
+		logger.Info("Deleting connection")
+		r.ConnPool.Delete(key)
+		if err := r.removeFinalizer(ctx, &logger, csiAddonsNode); err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+
+	if err := r.addFinalizer(ctx, &logger, csiAddonsNode); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	logger.Info("Connecting to sidecar")
+	newCon, err := connection.NewConnection(ctx, endPoint, nodeID, driverName)
+	if err != nil {
+		logger.Error(err, "Failed to establish connection with sidecar")
+
+		errMessage := util.GetErrorMessage(err)
+		csiAddonsNode.Status.State = csiaddonsv1alpha1.CSIAddonsNodeStateFailed
+		csiAddonsNode.Status.Message = fmt.Sprintf("Failed to establish connection with sidecar: %v", errMessage)
+		if statusErr := r.Client.Status().Update(ctx, csiAddonsNode); statusErr != nil {
```

Comment:
Still wondering about the `Update()` vs `Patch()` call

Reply:
> Still wondering about the `Update()` vs `Patch()` call

again most of the operators and controllers user `Update()`, I don't see any advantage of explicitly specifying what to patch when you can just update the entire object.
https://book.kubebuilder.io/cronjob-tutorial/controller-implementation.html#2-list-all-active-jobs-and-update-the-status

I didn't find any such examples of Patch() being used by controller/operator.

Reply:
common practice is to use Update for resource update ex:- 
https://github.com/operator-framework/operator-sdk-samples/blob/caf90347ff2e40ae9511d651d6fef4bb42c891af/go/memcached-operator/controllers/memcached_controller.go#L114-L118